### PR TITLE
Initial work on a VirtualStorage class.

### PR DESCRIPTION
**Pull request by jetuk**
Currently enforces the volume constraint in the solver.

Still needs to influence the costs of routes that are connected to it.

### CODE REVIEW

**Comment by jetuk** [comment]
I should say, the reason for doing this is to implement `Parameters` that have some sort of storage element to them. We have the current `Licence` classes but they effectively emulate a crude storage in Python space, so we can acknowledge this should be a core part of the LP.

It raised a different question about licences also. Should they be a hard or soft constraint on flow?

File: pywr/solvers/cython_glpk.pyx [hunk 1]
```diff
@@ -211,8 +218,27 @@ cdef class CythonGLPKSolver:
...
+                ind[1+n] = self.idx_col_routes+c
+                val[1+n] = -num
+
+            glp_set_mat_row(self.prob, self.idx_row_virtual_storages+col, len(cntr), ind, val)
+            free(ind)
+            free(val)
+
         # aggregated node flow ratio constraints
-        self.idx_row_aggregated = self.idx_row_storages + len(storages)
+        self.idx_row_aggregated = self.idx_row_storages + len(virtual_storages)
```
Comment (snorfalorpagus) [inline pywr/solvers/cython_glpk.pyx]: This should be:

```
self.idx_row_aggregated = self.idx_row_virtual_storages + len(virtual_storages)
```

Alternatively, we could set `self.idx_row_aggregated = None`, then in the loop when the first row is created (`if ... is None`) assign it the value returned by `glp_add_rows`.

**Comment by snorfalorpagus** [comment]
Licences should be a hard constraint. A node should not be allowed to breach it's licence.

The complexity arrises from annual (or seasonal/rolling) licences where the equivalent daily average (= volume / 365) is less than the daily licence. At this point you need to manage it as a finite resource. The way this is done will depend on the specifics of the system. Some sources will be operated at their daily equivalent rate all year long, while others are really just operate during the peak period.

I think in principle I'm not opposed to moving the constraint from Python-space to the LP.

**Comment by jetuk** [comment]
If we implement licences with a virtual storage then they won't be able to breach the conditions. However we will also want their cost to increase if they are relatively overused. Is the intention then that a licence parameter creates its own `VirtualStorage` to limit the total flow, but is designed to be assigned to a cost attribute?

**Comment by jetuk** [comment]
I'm going to rebase this PR since #190 merged.

Seems like all the licence parameters will need rewriting to use it.

**Comment by jetuk** [comment]
So this is working a bit better now. The tests fail only on lpsolve. I have a half written `RecessionCurve` class to complete.

File: pywr/_core.pyx [hunk 2]
```diff
@@ -500,7 +500,7 @@ cdef class AggregatedNode(AbstractNode):
             self.model.dirty = True
 
     cpdef after(self, Timestep ts):
-        AbstractStorage.after(self, ts)
+        AbstractNode.after(self, ts)
```
Comment (snorfalorpagus) [inline pywr/_core.pyx]: Ouch. Nasty bug. I wonder how this went unnoticed?

Comment (jetuk) [inline pywr/_core.pyx]: I think all the work was in `Storage.after` rather than `AbstractStorage.after`.

File: pywr/solvers/cython_glpk.pyx [hunk 3]
```diff
@@ -211,8 +216,37 @@ cdef class CythonGLPKSolver:
             free(ind)
             free(val)
 
+        # virtual storage
+        if len(virtual_storages):
+            self.idx_row_virtual_storages = glp_add_rows(self.prob, len(virtual_storages))
+        for col, storage in enumerate(virtual_storages):
+            # We need to handle the same route appearing twice here.
```
Comment (snorfalorpagus) [inline pywr/solvers/cython_glpk.pyx]: Is this a TODO?

File: pywr/solvers/cython_glpk.pyx [hunk 4]
```diff
@@ -211,8 +216,37 @@ cdef class CythonGLPKSolver:
...
+            cols = {}
```
Comment (snorfalorpagus) [inline pywr/solvers/cython_glpk.pyx]: I've avoided dictionaries elsewhere in the solver. Is this really needed?

File: pywr/solvers/cython_glpk.pyx (hunk 3 above)
Comment (jetuk) [inline pywr/solvers/cython_glpk.pyx]: No, it works with that.

File: pywr/solvers/cython_glpk.pyx (hunk 4 above)
Comment (jetuk) [inline pywr/solvers/cython_glpk.pyx]: It is needed to handle the multiple routes I think.

**Comment by jetuk** [comment]
What's left to add to this before merge? The licence refactor?

**Comment by snorfalorpagus** [comment]
Yes. Will this work OK with rolling licences? I guess the volume would need to be reset each timestep?

**Comment by snorfalorpagus** [comment]
Support for lpsolve is done. Need to add JSON support, and some documentation (esp. how the factors work).

**Comment by snorfalorpagus** [comment]
The virtual storage class has a "cost" parameter that I don't think is respected. I think this is just a carry-over from storage and should be removed? @jetuk

**Comment by jetuk** [comment]
That's true about cost. TBH the  hierarchy for the `Storage` classes probably needs a more fundamental think through. I.e. which should have which parameter.

**Comment by snorfalorpagus** [comment]
Let's talk about this tomorrow. I can see some sense in allowing a virtual storage to have a cost.

**Pull request merged**